### PR TITLE
Port param plugin

### PR DESCRIPTION
**Pull request by rob-clarke**
Initial work to port param plugin to ROS2
Compiles but not yet tested.

Progress for #1559

### CODE REVIEW

**Comment by vooon** [comment]
Oh, sorry, i done that port few hours ago. Not yet sure that it's working as i want.

I will review your port later. Maybe i could get some better ideas...
But i certainly wouldn't have time to deal with mission plugins till next weekend.

**Comment by vooon** [comment]
Also please note that i'm thinking to use futures instead of cond vars where i can.

File: mavros/src/plugins/param.cpp [hunk 1]
```diff
@@ -232,85 +194,52 @@ class Parameter
...
 
-    // XXX(vooon): find better solution
-    msg.header.stamp = ros::Time::now();
-
-    msg.param_id = param_id;
-    msg.value.integer = to_integer();
-    msg.value.real = to_real();
+    msg.param_id = get_name();
+    msg.value.integer = as_int();
+    msg.value.real = as_double();
```
Comment (vooon) [inline mavros/src/plugins/param.cpp:242]: At that point you'll get exception: one of as_*() would throw it.

File: mavros/src/plugins/param.cpp [hunk 2]
```diff
@@ -48,179 +54,135 @@ using utils::enum_value;
...
-        int_tmp = uv.param_uint32;
-        param_value = int_tmp;
-        break;
+      	return Parameter(param_name, uv.param_int32, pmsg.param_count, pmsg.param_index);
       case enum_value(MT::REAL32):
-        float_tmp = uv.param_float;
-        param_value = float_tmp;
-        break;
-      // [[[end]]] (checksum: 5950e4ee032d4aa198b953f56909e129)
+      	return Parameter(param_name, uv.param_float, pmsg.param_count, pmsg.param_index);
```
Comment (vooon) [inline mavros/src/plugins/param.cpp:103]: That part of conversions should work - ints are convertible to PARAMETER_INTEGER.

File: mavros/src/plugins/param.cpp [hunk 3]
```diff
@@ -48,179 +54,135 @@ using utils::enum_value;
...
-        int_tmp = pmsg.param_value;
-        param_value = int_tmp;
-        break;
+      	return Parameter(param_name, pmsg.param_value, pmsg.param_count, pmsg.param_index);
       case enum_value(MT::REAL32):
-        float_tmp = pmsg.param_value;
-        param_value = float_tmp;
-        break;
-      // [[[end]]] (checksum: c30ee34dd84213471690612ab49f1f73)
+      	return Parameter(param_name, pmsg.param_value, pmsg.param_count, pmsg.param_index);
```
Comment (vooon) [inline mavros/src/plugins/param.cpp:136]: But here you likely get wrong type: pmsg.param_value is float, but it carries bools and integers.

File: mavros/src/plugins/param.cpp [hunk 4]
```diff
@@ -365,37 +294,34 @@ class ParamSetOpt
...
+    push_srv = node->create_service<mavros_msgs::srv::ParamPush>("push", std::bind(&ParamPlugin::push_cb, this, _1, _2));
+    set_srv = node->create_service<mavros_msgs::srv::ParamSet>("set", std::bind(&ParamPlugin::set_cb, this, _1, _2));
+    get_srv = node->create_service<mavros_msgs::srv::ParamGet>("get", std::bind(&ParamPlugin::get_cb, this, _1, _2));
 
-    pull_srv = param_nh.advertiseService("pull", &ParamPlugin::pull_cb, this);
-    push_srv = param_nh.advertiseService("push", &ParamPlugin::push_cb, this);
-    set_srv = param_nh.advertiseService("set", &ParamPlugin::set_cb, this);
-    get_srv = param_nh.advertiseService("get", &ParamPlugin::get_cb, this);
+    auto parameters_qos = rclcpp::ParametersQoS();
+    param_value_pub = node->create_publisher<mavros_msgs::msg::Param>("param_value", parameters_qos);
```
Comment (vooon) [inline mavros/src/plugins/param.cpp:318]: I should use that qos.

File: mavros/src/plugins/param.cpp (hunk 3 above)
Comment (rob-clarke) [reply mavros/src/plugins/param.cpp:136]: I think my plan was to wrap it in an appropriate `type(pmsg.param_value)`  style cast to force the template. Evidently I never got around to it.

**Comment by rob-clarke** [comment]
Added some fixes for "completeness". As you've already got the port done I'll close this

**Pull request closed without merging**